### DISCUSSION TITLE
spi_read_blocking and spi_write_blocking and how to use it right

### DISCUSSION THREAD

**Discussion by kamikaravioli** [MrYsLab/telemetrix#58] [Q&A] [answered]
Hello,
first of all many thanks for the work done.  Like the title of the discussion reveals the topic is spi_read and write and how to use it the right way. 

I used telemetrix python API with an Read-only sensor and it worked flawlessly. Now I'm trying to read a more complex chip and I'm not sure if I can realise it with telemetrix. So I need more input to understand it better and estimate needed effort. 

I have a simple sketch/library to read a register from an device over SPI. It works perfectly in Arduino IDE. I'm trying to "rewrite" this very simple sketch to read/write a single register for ADS1256 to understand how telemetrix spi_write_blocking and spi_read_blocking works. 

So like I said the arduino sketch/library is very simple but it works fine:

```
#include <SPI.h>
const int CS_PIN = 10;

// Reading
long readRegister(uint8_t registerAddress)
{
SPI.beginTransaction(SPISettings(1920000, MSBFIRST, SPI_MODE1));
digitalWrite(CS_PIN, LOW);
SPI.transfer(0x10 | registerAddress);
SPI.transfer(0);
delayMicroseconds(5);
int _registerValuetoRead = SPI.transfer(0xFF);
digitalWrite(CS_PIN, HIGH);
SPI.endTransaction();
delay(100);
return _registerValuetoRead;
}

// Writing
void writeRegister(uint8_t registerAddress, uint8_t registerValueToWrite)
{
SPI.beginTransaction(SPISettings(1920000, MSBFIRST, SPI_MODE1));
digitalWrite(CS_PIN, LOW);
delayMicroseconds(5);
SPI.transfer(0x50 | registerAddress);
SPI.transfer(0);
SPI.transfer(registerValueToWrite);
digitalWrite(CS_PIN, HIGH);
SPI.endTransaction();
delay(100);
}


void setup() {
  SPI.begin();
  pinMode(CS_PIN, OUTPUT);
}

void loop() {
  int datarate = readRegister(0x03);
  writeRegister(0x03, 0b11110000);
}
```
It's not really important, I'm more interested in the implementation of the exact same function with telemetrix, but maybe its better for understanding what the sketch function does:
In this example I read the 3rd register of the ADS1256, that represents the value of the data rate. To read the DRATE-register on ADS1256 I have to send two command bytes to the device: 0001 rrrr and 0000 nnnn (rrrr = address of register, nnnn = number of bytes to read - 1). So to read DRATE we have to send "0001 0000 | 0000 0011" (= 0001 0011) and then 0000 0000 to read only one byte. To write we have to send:  0101 wwww, 0000 nnnn (wwww = address of the register, nnnn = is the number of bytes to write -1) and the desired value. To write new DRATE value to register: "0101 0000 | 0000 0011" (= 0101 0011), 0000 0000 and 11110000 (for decimal 240 that represents 30000 SPS). The read function returns "130" that represents SPS of 100. The write function writes value of 240 that represents SPS of 30000. Both works flawlessly in Arduino IDE. 

I tried to "translate"/implement the function with telemetrix, but not successful.
```
from telemetrix import telemetrix

board = telemetrix.Telemetrix()

CS = [10]
CS_PIN = 10

def print_function(report):
    print(report)

def read_register(register, number_bytes, callback_function):
    board.spi_cs_control(CS_PIN, 0)
    board.spi_write_blocking([0x10 | register, 0])
    time.sleep(.005)
    board.spi_read_blocking(register, number_bytes, call_back=callback_function)
    board.spi_cs_control(CS_PIN, 1)
	
board.set_pin_mode_spi(CS)
#board.spi_set_format(4, 1, 0x04)

while True:
    try:
        read_register(0x03, 1, print_function)
    except KeyboardInterrupt:
        board.shutdown()
        sys.exit(0)
```

Does my python/telemetrix script makes any sense (output that I get makes none)? Is it the way it should work or am I completely wrong? If we ignore the specific chip (ADS1256) and look only on our sketch function (readRegister for example), how would the telemetrix code for exact the same function look like?

Is it even possible to implement the functions from sketch with telemetrix or am I wasting my time?
Thank you.

**Comment by MrYsLab** [comment, marked as answer]
@kamikaravioli 
I genuinely apologize for getting back to you so late. The notification for this post was placed in my junk mailbox for some reason, and I just noticed it. 

Some changes will be coming to SPI support soon to be compatible with the latest versions of the Arduino SPI library and to allow one to specify the chip select pin when doing spi reads and writes.

Please let me know if you still want me to address your question. 

Again, I'm sorry for not getting back to you sooner.
